File: acwing/算法基础课/第五讲/5. 多重背包问题 II.py
N = 2010

f = [0] * N

n, m = map(int, input().split())


for i in range(1, n + 1):
    goods = []
    v, w, s = map(int, input().split())
    k = 1
    while k <= s:
        s -= k
        goods.append((k * v, k * w))
        k *= 2
    if s > 0:
        goods.append((s * v, s * w))
    for good in goods:
        for j in range(m, good[0] - 1, - 1):
            f[j] = max(f[j], f[j - good[0]] + good[1])

print(f[m])

File: acwing/算法基础课/第五讲/5.多重背包问题II/5.多重背包问题II.cpp
#include <iostream>
#include <algorithm>
#include <cstring>
#include <vector>

using namespace std;

const int N = 2010;

int f[N];

typedef pair<int, int> PII;

int main()
{
    int n, v;
    cin >> n >> v;
    for (int i = 1; i <= n; i++)
    {
        int vi, wi, si;
        cin >> vi >> wi >> si;
        vector<PII> goods;

        int k = 1;
        while (si >= k)
        {
            si -= k;
            goods.push_back(make_pair(vi * k, wi * k));
            k *= 2;
        }
        if (si > 0)
        {
            goods.push_back(make_pair(vi * si, wi * si));
        }
        for (auto good : goods)
        {
            for (int j = v; j >= good.first; j--)
            {
                f[j] = max(f[j], f[j - good.first] + good.second);
            }
        }
    }
    cout << f[v] << endl;
    return 0;
}